lib/Runtime/Library/LazyJSONString.cpp

```diff
@@ -0,0 +1,139 @@
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "RuntimeLibraryPch.h"
+
+namespace Js
+{
+
+LazyJSONString::LazyJSONString(_In_ JSONProperty* jsonContent, charcount_t length, _In_opt_ const char16* gap, charcount_t gapLength, _In_ StaticType* type) :
+    JavascriptString(type),
+    jsonContent(jsonContent),
+    gap(gap),
+    gapLength(gapLength)
+{
+    // Use SetLength to ensure length is valid
+    SetLength(length);
+}
+
+DynamicObject*
+LazyJSONString::ParseObject(_In_ JSONObject* valueList) const
+{
+    const uint elementCount = valueList->Count();
+    PropertyIndex requestedInlineSlotCapacity = static_cast<PropertyIndex>(elementCount);
+    DynamicObject* obj = this->GetLibrary()->CreateObject(true, requestedInlineSlotCapacity);
+    FOREACH_DLISTCOUNTED_ENTRY(JSONProperty, Recycler, entry, valueList)
+    {
+        Var propertyValue = Parse(&entry);
+        PropertyValueInfo info;
+        PropertyString* propertyString = PropertyString::TryFromVar(entry.propertyName);
+        if (!propertyString || !propertyString->TrySetPropertyFromCache(obj, propertyValue, this->GetScriptContext(), PropertyOperation_None, &info))
+        {
+            JavascriptOperators::SetProperty(obj, obj, entry.propertyRecord->GetPropertyId(), propertyValue, &info, this->GetScriptContext());
+        }
+    }
+    NEXT_DLISTCOUNTED_ENTRY;
+    return obj;
+}
+
+JavascriptArray*
+LazyJSONString::ParseArray(_In_ JSONArray* jsonArray) const
+{
+    const uint32 length = jsonArray->length;
+    JavascriptArray* arr = this->GetLibrary()->CreateArrayLiteral(length);
+    JSONProperty* prop = jsonArray->arr;
+    for (uint i = 0; i < length; ++i)
+    {
+        Var element = Parse(&prop[i]);
+        BOOL result = arr->SetItem(i, element, PropertyOperation_None);
+        Assert(result); // Setting item in an array we just allocated should always succeed
+    }
+    return arr;
+}
+
+Var
+LazyJSONString::Parse(_In_ JSONProperty* prop) const
+{
+    switch (prop->type)
+    {
+    case JSONContentType::Array:
+        return ParseArray(prop->arr);
+    case JSONContentType::Null:
+        return this->GetLibrary()->GetNull();
+    case JSONContentType::True:
+        return this->GetLibrary()->GetTrue();
+    case JSONContentType::False:
+        return this->GetLibrary()->GetFalse();
+    case JSONContentType::Number:
+        return prop->numericValue.value;
+    case JSONContentType::String:
+        return prop->stringValue;
+    case JSONContentType::Object:
+        return ParseObject(prop->obj);
+    default:
+        Assume(UNREACHED);
+        return nullptr;
+    }
+}
+
+Var
+LazyJSONString::Parse() const
+{
+    // If the gap is a non-space character, then parsing will be non-trivial transformation
+    for (charcount_t i = 0; i < this->gapLength; ++i)
```

Comment:
I understand the reason behind this check but I have a couple of questions:
1. If the gap is a non-space character, will JSON.parse ever succeed? If not, can we either throw a straight up syntax error, or maybe cache some position/token info and then throw?
2. If the check here fails, rather than just return null, I would eagerly serialize the string here and throw away the metadata since that's no longer needed (actually, GetSz should probably throw away the intermediate metadata anyway since we dont need to extend the lifetime of those objects). We could even just move the check to the constructor (although delay serialization till Parse is called)

Reply:
1. I was thinking you might be able to have something perverse like a gap of `{"a":"WHY"},` and so if you do serialize `{b:"Because I can!"}` you get serialized string of
````
{
{"a":"WHY"},{"b": "Because I can!"}
}
````

2a. I considered throwing away the metadata if this fails but was lazy and thought it wouldn't happen in practice.

2b. I was wary of throwing away the metadata in case people call parse multiple times on the same string (or need the string and then later call parse on it). But could be that isn't a realistic concern.

Reply:
Yuck at 1. For 2, its probably good to release the metadata as soon as we know they're useless so that we don't waste memory. Also, would be nice to have a AutoPtr setting a flag indicating whether the complex parse failed or not and cache that- if it failed, we can throw the syntax error without having to retry the parse

Reply:
I feel like failing here would be so rare that storing any additional data about it is a waste of memory.

However, I will throw away the metadata after flattening in case of complex gap, because that seems like a much more realistic scenario.